ipq60xx: add Cambium Networks XE3-4 BDF files

skramstad opened the pull request:
The board-2.bin's are extracted from firmware `6.6.0.1-r11`. 
They use `WLAN.HK.2.5.r4-00745-QCAHKSWPL_SILICONZ-1.550583.2.561602.1 v1`.
The BDF and firmware has never been updated.

**IPQ6018**
Extracted working copy from running OS.

**QCN9074**
This is a 5GHz and 6GHz combined radio. You are able to switch between
those GHz, but we are having an issue with detecting ac/n for 5GHz. 
6GHz is detected as AX/legacy.

Signed-off-by: Kristian Skramstad <kristian+github@83.no>

testuser7 commented:
> They use `bdwlan.b10-puma.bin` as BDF for IPQ6018.
> But it does not work with latest WLAN.HK firmware

Can you try to update regdb using this tool and see if it works: https://github.com/testuser7/ath_bdf_tool ?

skramstad commented:
> > They use `bdwlan.b10-puma.bin` as BDF for IPQ6018.
> > But it does not work with latest WLAN.HK firmware
> 
> Can you try to update regdb using this tool and see if it works: https://github.com/testuser7/ath_bdf_tool ?

Thanks. I'll have a look tomorrow. Cheers 👍

testuser7 commented:
QCN9074 BDF is missing variant in file name.

skramstad commented:
> QCN9074 BDF is missing variant in file name.

Yes, I saw that too now. I'll fix that. 
I did manage to update regdb from the newest firmware, and will do some more testing on the different firmware versions.

I have 1 issue too with IPQ6018, that after a boot/restart I have to set `iw reg set XX` again.

But there's something in the bdf for qcn9074, since you can toggle between 5GHz and 6GHz and only detecting legacy and AX. It should detect more than just AX for 5GHz.

skramstad commented:
@testuser7 tested with updated regdb on ipq6018 and qcn9074. For ipq6018, firmware `bdwlan.b10-puma.bin` is not working. On WLAN.HK.2.5 I get legacy/AX only. On WLAN.HK.2.6 and above it detects n/ac/ax etc but not able to start because of `qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received`. Not able to recover and reboots.

Using `bdwlan.b13.original` works with every WLAN.HK versions for ipq6018. No memory issue and no `daemon.err hostapd: Driver does not support configured HT capability [HT40*]` I get with `bdwlan.b10-puma.bin`.
 
On WLAN.HK.2.5 I can enable and disable radios fine, but with WLAN.HK.2.7 it will get an error `daemon.notice hostapd: nl80211: Failed to remove interface phy0-ap0 from bridge br-lan: Invalid argument`. But is able to recover and no reboot.

My experience with qcn9074 BDF is that it's probably broken or not updated to work with latest WLAN.HK. Selecting 5GHz and trying to enable the radio gives me `daemon.err hostapd: Driver does not support configured HT capability [HT40*]`. Enabling 6GHz is fine, but I can't verify that the radio is working since I don't have a 6GHz device.

So I need to figure out the qcn9074 issues. I pushed a new commit with correct variant.
I'll keep `bdwlan.b13.original` since the other one is a total mess. And `bdwlan.b13.original` is stable and works.

skramstad commented:
@testuser7 Do you happen to know if coldboot is broken for QCN9074? I have to do a second reboot to detect the radio...

testuser7 commented:
@skramstad I don't have this problem. So the file `/lib/firmware/ath11k/QCN9074/hw1.0/cal-pci-0000:01:00.0.bin` is not created after first start?

skramstad commented:
> @skramstad I don't have this problem. So the file `/lib/firmware/ath11k/QCN9074/hw1.0/cal-pci-0000:01:00.0.bin` is not created after first start?

You are correct. I've also noticed the board-2.bin has `chmod +x` but the board-2.bin in IPQ6018 folder does not. But that's something different and probably how it should be.

skramstad commented:
@testuser7 Got the error again after a power off and on again.

```
[   22.249260] ath11k_pci 0000:01:00.0: failed to wait board file download request: -110
[   22.249314] ath11k_pci 0000:01:00.0: qmi failed to load bdf file
[   22.256119] ath11k_pci 0000:01:00.0: failed to load board data file: -110
```
Yes, the board-2.bin is in the folder.

testuser7 commented:
So this is ath11k driver related. It's failing due to timeout: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/net/wireless/ath/ath11k/qmi.c?h=v6.6.32#n2368
You can increase the timeout and check if it's working: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/net/wireless/ath/ath11k/qmi.h?h=v6.6.32#n14

skramstad commented:
Thanks. I'll have a look at the timeout.

skramstad commented:
> So this is ath11k driver related. It's failing due to timeout: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/net/wireless/ath/ath11k/qmi.c?h=v6.6.32#n2368 You can increase the timeout and check if it's working: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/net/wireless/ath/ath11k/qmi.h?h=v6.6.32#n14

Hmm... one question tho. Is it failing because it was unable to extract the calibration? Because I've seen on some boots that it's unable to extract. And when I did a second restart with `chmod -x` on my board-2.bin it's all good.
This is just a thought, but curious what you think.

testuser7 commented:
I'm not sure if this error is generated by a missing BDF or something else. You can check the code here: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/soc/qcom/qmi_interface.c#n292

skramstad commented:
@testuser7 I believe my qcn9074 BDF is broken in one way or another. If I try to enable 5GHz I get this message: `daemon.err hostapd: Driver does not support configured HT capability [HT40*]`.

```
Wiphy phy2
	wiphy index: 2
	max # scan SSIDs: 16
	max scan IEs length: 184 bytes
	max # sched scan SSIDs: 0
	max # match sets: 0
	Retry short limit: 7
	Retry long limit: 4
	Coverage class: 0 (up to 0m)
	Device supports RSN-IBSS.
	Device supports AP-side u-APSD.
	Supported Ciphers:
		* TKIP (00-0f-ac:2)
		* CCMP-128 (00-0f-ac:4)
		* CMAC (00-0f-ac:6)
		* CMAC-256 (00-0f-ac:13)
		* GMAC-128 (00-0f-ac:11)
		* GMAC-256 (00-0f-ac:12)
		* GCMP-128 (00-0f-ac:8)
		* GCMP-256 (00-0f-ac:9)
		* CCMP-256 (00-0f-ac:10)
	Available Antennas: TX 0xf RX 0xf
	Configured Antennas: TX 0xf RX 0xf
	Supported interface modes:
		 * managed
		 * AP
		 * monitor
		 * mesh point
	Band 2:
		HE Iftypes: managed
			HE MAC Capabilities (0x000b9a181040):
				+HTC HE Supported
				TWT Requester
				Dynamic BA Fragementation Level: 1
				BSR
				Broadcast TWT
				OM Control
				Maximum A-MPDU Length Exponent: 3
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c604c89ffdb839c110c00):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				Full Bandwidth UL MU-MIMO
				DCM Max Constellation: 1
				DCM Max Constellation Rx: 1
				SU Beamformer
				SU Beamformee
				MU Beamformer
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Ng = 16 MU Feedback
				Codebook Size SU Feedback
				Codebook Size MU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				STBC Rx > 80MHz
				HE ER SU PPDU 4x HE-LTF 0.8us GI
				HE ER SU PPDU 1x HE-LTF 0.8us GI
				TX 1024-QAM
				RX 1024-QAM
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: managed
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		HE Iftypes: AP
			HE MAC Capabilities (0x000d9a181040):
				+HTC HE Supported
				TWT Responder
				Dynamic BA Fragementation Level: 1
				BSR
				Broadcast TWT
				OM Control
				Maximum A-MPDU Length Exponent: 3
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c604c88ffdb839c110c00):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				Full Bandwidth UL MU-MIMO
				DCM Max Constellation Rx: 1
				SU Beamformer
				SU Beamformee
				MU Beamformer
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Ng = 16 MU Feedback
				Codebook Size SU Feedback
				Codebook Size MU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				STBC Rx > 80MHz
				HE ER SU PPDU 4x HE-LTF 0.8us GI
				HE ER SU PPDU 1x HE-LTF 0.8us GI
				TX 1024-QAM
				RX 1024-QAM
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: AP
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		HE Iftypes: mesh point
			HE MAC Capabilities (0x00098a081040):
				+HTC HE Supported
				Dynamic BA Fragementation Level: 1
				BSR
				OM Control
				Maximum A-MPDU Length Exponent: 1
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c600c80fd5b811c100000):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				SU Beamformer
				SU Beamformee
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Codebook Size SU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				HE ER SU PPDU 1x HE-LTF 0.8us GI
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: mesh point
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		Bitrates (non-HT):
			* 6.0 Mbps
			* 9.0 Mbps
			* 12.0 Mbps
			* 18.0 Mbps
			* 24.0 Mbps
			* 36.0 Mbps
			* 48.0 Mbps
			* 54.0 Mbps
		Frequencies:
			* 5180 MHz [36] (30.0 dBm)
			* 5200 MHz [40] (30.0 dBm)
			* 5220 MHz [44] (30.0 dBm)
			* 5240 MHz [48] (30.0 dBm)
			* 5260 MHz [52] (30.0 dBm) (radar detection)
			* 5280 MHz [56] (30.0 dBm) (radar detection)
			* 5300 MHz [60] (30.0 dBm) (radar detection)
			* 5320 MHz [64] (30.0 dBm) (radar detection)
			* 5500 MHz [100] (30.0 dBm) (radar detection)
			* 5520 MHz [104] (30.0 dBm) (radar detection)
			* 5540 MHz [108] (30.0 dBm) (radar detection)
			* 5560 MHz [112] (30.0 dBm) (radar detection)
			* 5580 MHz [116] (30.0 dBm) (radar detection)
			* 5600 MHz [120] (30.0 dBm) (radar detection)
			* 5620 MHz [124] (30.0 dBm) (radar detection)
			* 5640 MHz [128] (30.0 dBm) (radar detection)
			* 5660 MHz [132] (30.0 dBm) (radar detection)
			* 5680 MHz [136] (30.0 dBm) (radar detection)
			* 5700 MHz [140] (30.0 dBm) (radar detection)
			* 5720 MHz [144] (disabled)
			* 5745 MHz [149] (30.0 dBm) (radar detection)
			* 5765 MHz [153] (30.0 dBm) (radar detection)
			* 5785 MHz [157] (30.0 dBm) (radar detection)
			* 5805 MHz [161] (30.0 dBm) (radar detection)
			* 5825 MHz [165] (30.0 dBm) (radar detection)
			* 5845 MHz [169] (30.0 dBm) (radar detection)
			* 5865 MHz [173] (30.0 dBm) (radar detection)
			* 5885 MHz [177] (disabled)
	Band 4:
		HE Iftypes: managed
			HE MAC Capabilities (0x000b9a181040):
				+HTC HE Supported
				TWT Requester
				Dynamic BA Fragementation Level: 1
				BSR
				Broadcast TWT
				OM Control
				Maximum A-MPDU Length Exponent: 3
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c604c89ffdb839c110c00):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				Full Bandwidth UL MU-MIMO
				DCM Max Constellation: 1
				DCM Max Constellation Rx: 1
				SU Beamformer
				SU Beamformee
				MU Beamformer
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Ng = 16 MU Feedback
				Codebook Size SU Feedback
				Codebook Size MU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				STBC Rx > 80MHz
				HE ER SU PPDU 4x HE-LTF 0.8us GI
				HE ER SU PPDU 1x HE-LTF 0.8us GI
				TX 1024-QAM
				RX 1024-QAM
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: managed
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		HE Iftypes: AP
			HE MAC Capabilities (0x000d9a181040):
				+HTC HE Supported
				TWT Responder
				Dynamic BA Fragementation Level: 1
				BSR
				Broadcast TWT
				OM Control
				Maximum A-MPDU Length Exponent: 3
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c604c88ffdb839c110c00):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				Full Bandwidth UL MU-MIMO
				DCM Max Constellation Rx: 1
				SU Beamformer
				SU Beamformee
				MU Beamformer
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Ng = 16 MU Feedback
				Codebook Size SU Feedback
				Codebook Size MU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				STBC Rx > 80MHz
				HE ER SU PPDU 4x HE-LTF 0.8us GI
				HE ER SU PPDU 1x HE-LTF 0.8us GI
				TX 1024-QAM
				RX 1024-QAM
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: AP
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		HE Iftypes: mesh point
			HE MAC Capabilities (0x00098a081040):
				+HTC HE Supported
				Dynamic BA Fragementation Level: 1
				BSR
				OM Control
				Maximum A-MPDU Length Exponent: 1
				RX Control Frame to MultiBSS
				A-MSDU in A-MPDU
				OM Control UL MU Data Disable RX
			HE PHY Capabilities: (0x1c600c80fd5b811c100000):
				HE40/HE80/5GHz
				HE160/5GHz
				HE160/HE80+80/5GHz
				LDPC Coding in Payload
				HE SU PPDU with 1x HE-LTF and 0.8us GI
				STBC Tx <= 80MHz
				STBC Rx <= 80MHz
				SU Beamformer
				SU Beamformee
				Beamformee STS <= 80Mhz: 7
				Beamformee STS > 80Mhz: 7
				Sounding Dimensions <= 80Mhz: 3
				Sounding Dimensions > 80Mhz: 3
				Ng = 16 SU Feedback
				Codebook Size SU Feedback
				PPE Threshold Present
				HE SU PPDU & HE PPDU 4x HE-LTF 0.8us GI
				Max NC: 3
				HE ER SU PPDU 1x HE-LTF 0.8us GI
			HE RX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set <= 80 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE RX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
			HE TX MCS and NSS set 160 MHz
				1 streams: MCS 0-11
				2 streams: MCS 0-11
				3 streams: MCS 0-11
				4 streams: MCS 0-11
				5 streams: not supported
				6 streams: not supported
				7 streams: not supported
				8 streams: not supported
		EHT Iftypes: mesh point
			EHT MAC Capabilities (0x0000):
			EHT PHY Capabilities: (0x0000000000000000):
			EHT MCS/NSS: (0x):
			EHT bw <= 80 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw <= 80 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 8-9: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 10-11: Rx=0, Tx=0
			EHT bw=160 MHz, max NSS for MCS 12-13: Rx=0, Tx=0
		Bitrates (non-HT):
			* 6.0 Mbps
			* 9.0 Mbps
			* 12.0 Mbps
			* 18.0 Mbps
			* 24.0 Mbps
			* 36.0 Mbps
			* 48.0 Mbps
			* 54.0 Mbps
		Frequencies:
			* 5955 MHz [1] (30.0 dBm)
			* 5975 MHz [5] (30.0 dBm)
			* 5995 MHz [9] (30.0 dBm)
			* 6015 MHz [13] (30.0 dBm)
			* 6035 MHz [17] (30.0 dBm)
			* 6055 MHz [21] (30.0 dBm)
			* 6075 MHz [25] (30.0 dBm)
			* 6095 MHz [29] (30.0 dBm)
			* 6115 MHz [33] (30.0 dBm)
			* 6135 MHz [37] (30.0 dBm)
			* 6155 MHz [41] (30.0 dBm)
			* 6175 MHz [45] (30.0 dBm)
			* 6195 MHz [49] (30.0 dBm)
			* 6215 MHz [53] (30.0 dBm)
			* 6235 MHz [57] (30.0 dBm)
			* 6255 MHz [61] (30.0 dBm)
			* 6275 MHz [65] (30.0 dBm)
			* 6295 MHz [69] (30.0 dBm)
			* 6315 MHz [73] (30.0 dBm)
			* 6335 MHz [77] (30.0 dBm)
			* 6355 MHz [81] (30.0 dBm)
			* 6375 MHz [85] (30.0 dBm)
			* 6395 MHz [89] (30.0 dBm)
			* 6415 MHz [93] (30.0 dBm)
			* 6435 MHz [97] (disabled)
			* 6455 MHz [101] (disabled)
			* 6475 MHz [105] (disabled)
			* 6495 MHz [109] (disabled)
			* 6515 MHz [113] (disabled)
			* 6535 MHz [117] (disabled)
			* 6555 MHz [121] (disabled)
			* 6575 MHz [125] (disabled)
			* 6595 MHz [129] (disabled)
			* 6615 MHz [133] (disabled)
			* 6635 MHz [137] (disabled)
			* 6655 MHz [141] (disabled)
			* 6675 MHz [145] (disabled)
			* 6695 MHz [149] (disabled)
			* 6715 MHz [153] (disabled)
			* 6735 MHz [157] (disabled)
			* 6755 MHz [161] (disabled)
			* 6775 MHz [165] (disabled)
			* 6795 MHz [169] (disabled)
			* 6815 MHz [173] (disabled)
			* 6835 MHz [177] (disabled)
			* 6855 MHz [181] (disabled)
			* 6875 MHz [185] (disabled)
			* 6895 MHz [189] (disabled)
			* 6915 MHz [193] (disabled)
			* 6935 MHz [197] (disabled)
			* 6955 MHz [201] (disabled)
			* 6975 MHz [205] (disabled)
			* 6995 MHz [209] (disabled)
			* 7015 MHz [213] (disabled)
			* 7035 MHz [217] (disabled)
			* 7055 MHz [221] (disabled)
			* 7075 MHz [225] (disabled)
			* 7095 MHz [229] (disabled)
			* 7115 MHz [233] (disabled)
			* 5935 MHz [2] (disabled)
	Supported commands:
		 * new_interface
		 * set_interface
		 * new_key
		 * start_ap
		 * new_station
		 * new_mpath
		 * set_mesh_config
		 * set_bss
		 * authenticate
		 * associate
		 * deauthenticate
		 * disassociate
		 * join_ibss
		 * join_mesh
		 * remain_on_channel
		 * set_tx_bitrate_mask
		 * frame
		 * frame_wait_cancel
		 * set_wiphy_netns
		 * set_channel
		 * probe_client
		 * set_noack_map
		 * register_beacons
		 * start_p2p_device
		 * set_mcast_rate
		 * connect
		 * disconnect
		 * channel_switch
		 * set_qos_map
		 * set_multicast_to_unicast
	software interface modes (can always be added):
		 * monitor
	valid interface combinations:
		 * #{ managed } <= 1, #{ AP, mesh point } <= 16,
		   total <= 16, #channels <= 1, STA/AP BI must match, radar detect widths: { 20 MHz (no HT), 20 MHz, 40 MHz, 80 MHz, 80+80 MHz, 160 MHz }

	HT Capability overrides:
		 * MCS: ff ff ff ff ff ff ff ff ff ff
		 * maximum A-MSDU length
		 * supported channel width
		 * short GI for 40 MHz
		 * max A-MPDU length exponent
		 * min MPDU start spacing
	Device supports TX status socket option.
	Device supports HT-IBSS.
	Device supports SAE with AUTHENTICATE command
	Device supports scan flush.
	Device supports AP scan.
	Device supports per-vif TX power setting
	Driver supports full state transitions for AP/GO clients
	Driver supports a userspace MPM
	Driver/device bandwidth changes during BSS lifetime (AP/GO mode)
	Device supports static SMPS
	Device supports dynamic SMPS
	Device supports configuring vdev MAC-addr on create.
	Device supports randomizing MAC-addr in scans.
	max # scan plans: 1
	max scan plan interval: -1
	max scan plan iterations: 0
	Supported TX frame types:
		 * IBSS: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * managed: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP/VLAN: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * mesh point: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-client: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-GO: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-device: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
	Supported RX frame types:
		 * IBSS: 0x40 0xb0 0xc0 0xd0
		 * managed: 0x40 0xb0 0xd0
		 * AP: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * AP/VLAN: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * mesh point: 0xb0 0xc0 0xd0
		 * P2P-client: 0x40 0xd0
		 * P2P-GO: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * P2P-device: 0x40 0xd0
	Maximum associated stations in AP mode: 512
	Supported extended features:
		* [ RRM ]: RRM
		* [ SET_SCAN_DWELL ]: scan dwell setting
		* [ FILS_STA ]: STA FILS (Fast Initial Link Setup)
		* [ CQM_RSSI_LIST ]: multiple CQM_RSSI_THOLD records
		* [ CONTROL_PORT_OVER_NL80211 ]: control port over nl80211
		* [ ACK_SIGNAL_SUPPORT ]: ack signal level support
		* [ TXQS ]: FQ-CoDel-enabled intermediate TXQs
		* [ ENABLE_FTM_RESPONDER ]: enable FTM (Fine Time Measurement) responder
		* [ STA_TX_PWR ]: TX power control per station
		* [ CONTROL_PORT_NO_PREAUTH ]: disable pre-auth over nl80211 control port support
		* [ SCAN_FREQ_KHZ ]: scan on kHz frequency support
		* [ CONTROL_PORT_OVER_NL80211_TX_STATUS ]: tx status for nl80211 control port support
		* [ FILS_DISCOVERY ]: FILS discovery frame transmission support
		* [ UNSOL_BCAST_PROBE_RESP ]: unsolicated broadcast probe response transmission support
		* [ BSS_COLOR ]: BSS coloring support
```
And by looking at the channels I'm stuck at 20MHz.
```
Band 2:
	* 5180 MHz [36]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5200 MHz [40]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5220 MHz [44]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5240 MHz [48]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5260 MHz [52]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5280 MHz [56]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5300 MHz [60]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5320 MHz [64]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5500 MHz [100]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5520 MHz [104]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5540 MHz [108]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5560 MHz [112]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5580 MHz [116]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5600 MHz [120]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 600000 ms
	* 5620 MHz [124]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 600000 ms
	* 5640 MHz [128]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 600000 ms
	* 5660 MHz [132]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5680 MHz [136]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5700 MHz [140]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5720 MHz [144] (disabled)
	* 5745 MHz [149]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5765 MHz [153]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5785 MHz [157]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5805 MHz [161]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5825 MHz [165]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5845 MHz [169]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5865 MHz [173]
	  Maximum TX power: 30.0 dBm
	  Radar detection
	  Channel widths: 20MHz
	  DFS state: usable (for 267 sec)
	  DFS CAC time: 60000 ms
	* 5885 MHz [177] (disabled)
Band 4:
	* 5955 MHz [1]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5975 MHz [5]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 5995 MHz [9]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6015 MHz [13]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6035 MHz [17]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6055 MHz [21]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6075 MHz [25]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6095 MHz [29]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6115 MHz [33]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6135 MHz [37]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6155 MHz [41]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6175 MHz [45]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6195 MHz [49]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6215 MHz [53]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6235 MHz [57]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6255 MHz [61]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6275 MHz [65]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6295 MHz [69]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6315 MHz [73]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6335 MHz [77]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6355 MHz [81]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6375 MHz [85]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6395 MHz [89]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6415 MHz [93]
	  Maximum TX power: 30.0 dBm
	  Channel widths: 20MHz
	* 6435 MHz [97] (disabled)
	* 6455 MHz [101] (disabled)
	* 6475 MHz [105] (disabled)
	* 6495 MHz [109] (disabled)
	* 6515 MHz [113] (disabled)
	* 6535 MHz [117] (disabled)
	* 6555 MHz [121] (disabled)
	* 6575 MHz [125] (disabled)
	* 6595 MHz [129] (disabled)
	* 6615 MHz [133] (disabled)
	* 6635 MHz [137] (disabled)
	* 6655 MHz [141] (disabled)
	* 6675 MHz [145] (disabled)
	* 6695 MHz [149] (disabled)
	* 6715 MHz [153] (disabled)
	* 6735 MHz [157] (disabled)
	* 6755 MHz [161] (disabled)
	* 6775 MHz [165] (disabled)
	* 6795 MHz [169] (disabled)
	* 6815 MHz [173] (disabled)
	* 6835 MHz [177] (disabled)
	* 6855 MHz [181] (disabled)
	* 6875 MHz [185] (disabled)
	* 6895 MHz [189] (disabled)
	* 6915 MHz [193] (disabled)
	* 6935 MHz [197] (disabled)
	* 6955 MHz [201] (disabled)
	* 6975 MHz [205] (disabled)
	* 6995 MHz [209] (disabled)
	* 7015 MHz [213] (disabled)
	* 7035 MHz [217] (disabled)
	* 7055 MHz [221] (disabled)
	* 7075 MHz [225] (disabled)
	* 7095 MHz [229] (disabled)
	* 7115 MHz [233] (disabled)
	* 5935 MHz [2] (disabled)
```
In Luci it shows AX/legacy mode only. So I'm missing ac/n etc...

Trying to downgrade WLAN.HK firmwares, and even testing the stock firmware don't do anything. So I dunno how they are able to make it work on stock software.
In Luci it show 2 bands tho, but I can't see that in `iw`. Is it something in the `mac80211.sh` script or iw or hostapd?

I'm lost here, but reading the ath11k mailing list shows a similar issue. But they talked about board-id, bad radio cards etc... I did try to use board-id 160 to 164 + no board-id but those did not load.

testuser7 commented:
Are you able to start ap on 5G and 6G with option HT20 and connect?

skramstad commented:
> Are you able to start ap on 5G and 6G with option HT20 and connect?

5G = NO
6G = Starts up but not detected

About 5G, if it matters, in Luci I see that the wifi channel is stuck on 6.123 -something. Does not change if I change channel.

skramstad commented:
@testuser7 so I got a little further in terms of testing BDF. I was able to get the qcn9074 detected with correct channels, ax/ax/n etc, with the qca BDF `bus=pci,qmi-chip-id=0,qmi-board-id=160.bin` from WLAN.HK.2.7. 

But then again, I loose 6GHz. I also tried to downgrade the regdb on this BDF with the one from stock that has 5 and 6GHz enabled. But that did nothing.

So if I'm able to enable 6GHz in the `bus=pci,qmi-chip-id=0,qmi-board-id=160.bin` then I could do some more testing. But I guess this is not possible.

testuser7 commented:
Can you try to use `board-id=162` for 6GHz?

skramstad commented:
> Can you try to use `board-id=162` for 6GHz?

I did earlier. But 160 was the only one that did load. 162 crashed.

testuser7 commented:
In my case, I can run the 5G radio using the board file with id `160`. The real power is very low, but it works.
Using the board file with id `162`, I can get the 6G radio to work, but I still don't see it on the client.

skramstad commented:
So the `board-id` is related to band? I did read this https://lore.kernel.org/all/MW2PR02MB3899226DC8F4EED0BF9A66ABED3C9@MW2PR02MB3899.namprd02.prod.outlook.com/T/ and is has some interesting information.

testuser7 commented:
No, everything is inside BDF. The board file with id `162` is prepared for the 6G band.

skramstad commented:
I was able to load `162` if I removed the caldata. But the result in i.e Luci was that I had only 1 channel and `dBm/mW` was way off. In the qca DTS they use `board_id = <0xa4>;`. But I guess that doesn't do anything. 
In my very basic research comparing the `qca 160` and `stock 255` with hexdump, I saw a gap in the qca file, that the stock had filled. Maybe that was the 6GHz code... But I guess we need `bdf2txt.py` and `txt2bdf.py` to edit.

@Ansuel I'm wondering if there's anything you are able to do with a BDF to make it more "compatible" in terms of newer support. This is for qcn9074 that can toggle between 5GHz and 6GHz.

testuser7 commented:
Did you try to use board file with id `164`?

skramstad commented:
Yeah. But no success. Do you have 2 bands on your qcn9074 too. 5+6GHz that you can toggle?

testuser7 commented:
No, only 6GHz. But I can start AP on 5GHz band by providing other BDF but the real power is very very low than.

skramstad commented:
Ok. I'll wait to see the respons from Ansuel.
Thanks for all the help @testuser7 👍

skramstad commented:
Just did some 6GHz testing. And it doesn't do anything. So the BDF for qcn9074 is a total mess. We do have a working 5GHz radio with another BDF (160) as backup, if we can't fix the stock BDF for 6GHz.

testuser7 commented:
My 6G radio works. The problem was with the Intel client. Can you share the firmware update file?

skramstad commented:
I've tested all the available BDF's for QCN9074. I now updated with the one they use while running. I also updated regdb for this file from WLAN.HK.2.9. 
However, I'm wondering if this `cnss2.bdf_pci0=0xab` is a clue.

All they do is
```
elif [ $SKU -eq $sku_jaguar_puma ]; then
# For Puma, use b10 for 2.4GHz and 5GHz radios
install -D ${CFDIR}/bdwlan.b10-puma ${CTDIR}/bdwlan.bin
install -D ${CFDIR}/bdwlan.b10-puma ${CTDIR}/bdwlan.b13
# 6GHz BDF file is different as compared to Lion
install -D ${QFDIR}/bdwlan.bab-puma ${QTDIR}/bdwlan.bin
install -D ${QFDIR}/bdwlan.bab-puma ${QTDIR}/bdwlan.bab
```

skramstad commented:
@testuser7 Do you know why I get `Regdomain is not set` with command `ath11k_bdf_tool -r bdwlan.bab`? Happens on both of the BDF's in this PR. Also tried on a qca OEM BDF with same warning.

testuser7 commented:
> @testuser7 Do you know why I get `Regdomain is not set` with command `ath11k_bdf_tool -r bdwlan.bab`? Happens on both of the BDF's in this PR. Also tried on a qca OEM BDF with same warning.
> 

If regdomain is not set, you do not need to remove them. You can read more about this here: https://fcc.report/FCC-ID/B94HHI31C/1616417.pdf

> Edit: Sorry, if it's set, a country code is set as default when BDF is loaded? Like US. So for me it's not set so that's why the warning is displayed?
> 

If the regdomain is set, you cannot change the country: https://forum.openwrt.org/t/adding-openwrt-support-for-linksys-mx5300/92667/36

> Edit2: Do you or anyone knows why the country is always set to US after boot. It does not respect Luci setting. Like it's overriding my selection in Luci...
> 
> ```
> phy#1 (self-managed)
> country US: DFS-FCC
> ```

What is your wireless config `/etc/config/wireless`?

skramstad commented:
@testuser7 
```
option country 'NO'
```

testuser7 commented:
It appears that `bdwlan.bab` is the only variant that should be used on 5G/6G radios. Try to find the document on Google using the keyword `"bdwlan.bab"`. Maybe you'll find something interesting there.

skramstad commented:
@robimarko is it important to add reserved-memory for qcn9000/qcn9074?
```
qcn9000_pcie0@50200000 {
	no-map;
	reg = <0x00 0x50200000 0x00 0x2d00000>;
};
```

skramstad commented:
@testuser7 the change that made the BDF for ipq6018 work was:
```
1c1
< 00000000: 0100 0404 0000 0000 0000 a632 0101 0003  ...........2....
---
> 00000000: 0100 0404 0000 0000 0000 a633 0101 0003  ...........3....
32c32
< 000001f0: 0300 0000 0001 0000 0000 0000 0000 0000  ................
---
> 000001f0: 0300 0000 0000 0000 0000 0000 0000 0000  ................
```
But it bothers me that they from early on they shipped the working BDF for ipq6018, but in the latest versions they changed it to a non working.

testuser7 commented:
This is the only difference between `bdwlan.b13.original` and `bdwlan.b10-puma.bin`?

skramstad commented:
> This is the only difference between `bdwlan.b13.original` and `bdwlan.b10-puma.bin`?

The diff is from `bdwlan.bin` coming from an early firmware, compared to `bdwlan.b10-puma.bin`.
Those 2 changes enabled ax/ac/n/a etc compared to ax/a only.

testuser7 commented:
The correlation between these files is not clear to me...

skramstad commented:
> The correlation between these files is not clear to me...

Forget about the `bdwlan.b13.original` since that was my replacement firmware as the `bdwlan.b10-puma.bin` failed. 
But, they copy `bdwlan.b10-puma.bin` from `WIFI_FW` folder and rename it to `bdwlan.bin` and put it in `/tmp/ipq6018/bdwlan.bin`. Since the latest `bdwlan.b10-puma.bin` did not work, I started testing every firmware versions they released, and the earliest version contains the one that is working, named bdwlan.bin from the `/tmp/ipq6018/bdwlan.bin`.

I then compared those two BDF's, and the result is what I posted earlier.

Hope this helps. But ipq6018 is OK now, so no worries.

In terms of qcn9074 BDF, I've tested other brands that also use a software based toggle between 5GHz and 6GHz. The Zyxel nwa220ax-6e has this feature as well. I extracted BDF and tested but same issue as with my BDF. I'm thinking the support for this "selectable" band feature is not present at the moment. I've also tested 8devices Pine BDF but no luck.

So if Ansuel and/or Robi accepts a working 5GHz radio since 6GHz is not working, then we can move on, an get this merged. I just have to upload the working BDF for QCN9074.

skramstad commented:
Hi @Ansuel, are you able to, or have some time to have a look at the `board-cambiumnetworks_xe34.qcn9074`? It's a dual 5GHz/6GHz single band BDF, so not sure if they are different from previous BDF you gave some love 👍

robimarko commented:
> @robimarko is it important to add reserved-memory for qcn9000/qcn9074?
> 
> ```
> qcn9000_pcie0@50200000 {
> 	no-map;
> 	reg = <0x00 0x50200000 0x00 0x2d00000>;
> };
> ```

No, ath11k will get the memory assigned by the kernel.

robimarko commented:
If IPQ6018 BDF is working fine as well as 5GHz on QCN9074 that is fine by me since the 6GHz BDF-s are mistery to me and QCA separated them

skramstad commented:
> If IPQ6018 BDF is working fine as well as 5GHz on QCN9074 that is fine by me since the 6GHz BDF-s are mistery to me and QCA separated them

Nice. I'll commit the new BDF for QCN9074 soon.

skramstad commented:
New BDF for QCN9074 is pushed.

robimarko commented:
Thanks! Rebased on top of main and merged!
